Add dispersion relation for MHD waves

TrestanSimon opened the pull request:
## Description

>Add functionality for the dispersion relations for Alfvén, slow magnetosonic, and fast magnentosonic waves.

## Motivation and context

From  https://github.com/PlasmaPy/PlasmaPy/issues/2197:
> We currently have some two-fluid dispersion relation solvers, but we don't have anything for plain MHD waves. Adding functionality for MHD wave dispersion relations would be particularly useful as a tool for pedagogy, and would also be useful as a limiting case for researchers.

## Implementation

There are three separate classes for each wave mode (`AlfvenWave`, `FastMagnetosonicWave`, and `SlowMagnetosonicWave`) which all derive from an abstract base class `AbstractMHDWave`. Each wave mode class overrides the abstract method `angular_frequency` from the base class. The associated constructors have required arguments for $B$, the ion species, and the ion number density* with optional arguments for the temperatures and adiabatic indices of the electrons and ions and the mean ionization state.

*Along with the wavenumber and angle of propagation provided when calling the `angular_frequency` method, these are the only parameters needed for calculating the angular frequency of the Alfvén wave and of the fast and slow magnetosonic waves in the cold plasma limit. (The temperatures will default to zero.)

Note also that some input validation code has already been written for [plasmapy/dispersion/analytical/two_fluid_.py](https://github.com/PlasmaPy/PlasmaPy/blob/e125e69c15dacbbaa6509c8d5c274d7c4d416374/plasmapy/dispersion/analytical/two_fluid_.py), so some of it is reused here.

## Related issues

Closes https://github.com/PlasmaPy/PlasmaPy/issues/2197
See also https://github.com/PlasmaPy/PlasmaPy/issues/2198

namurphy reviewed (changes requested):
@TrestanSimon — thank you for submitting this!  I'm excited to have it be added. 🚀 

I have a few suggested refinements (mostly related to implementing `@particle_input`), which I meant to post earlier but was traveling.

Thank you again! 😸

namurphy commented on plasmapy/dispersion/analytical/mhd_wave_class.py:
```diff
@@ -0,0 +1,262 @@
+    """
+
+    def angular_frequency(self, k: u.rad / u.m, theta: u.rad):
+        """
```
```suggestion
        r"""
```

namurphy commented on plasmapy/dispersion/analytical/tests/test_mhd_wave_class.py:
```diff
@@ -0,0 +1,101 @@
+                    "slow": [0, 0] * u.m / u.s,
+                },
+            ),
+            (  # No magnetic field
```
Good idea to test this limit!

TrestanSimon replied:
Thanks. I actually have a question about this particular limit. Without a magnetic field, the MHD fast mode would reduce to a "sound wave" regardless of the ratio between $T_e$ and $T_i$; however, from the two-fluid model, ion acoustic waves could only exist if $T_e \gg T_i$. Do you think some sort of warning should be added?

namurphy replied:
Good question!  Usually for MHD, it's assumed that the ion & electron temperatures are the same.  Most of the references I've found assume that the sound speed is given by the ideal relation

$$ c_s ≡ \frac{γ k_B T}{m_i} $$

(with $γ = \frac{5}{3}$) which I believe should be used instead of the ion sound speed used for ion acoustic waves. I'll double check this with some friends who work with plasma waves a lot more than I do.  If we assume the same temperature for ions & electrons, we would remove the need for that warning.

namurphy reviewed (commented):
Thank you for working on this!  I have a few minor suggestions now, and will hopefully have a chance to look through this in a bit more detail later this week.

Also — sorry about the delay!  I was traveling last week. ✈️

TrestanSimon commented:
@namurphy Thanks for all of your suggestions.

I am unfamiliar with how docstring inheritance works between child and parent classes and for overridden methods with Sphinx. I have had trouble finding consistent info about this online and have not seen any examples of it within PlasmaPy. Is there some decorator I should use for this? Or would something like `__doc__ += ...` work?

namurphy commented:
> I am unfamiliar with how docstring inheritance works between child and parent classes and for overridden methods with Sphinx. I have had trouble finding consistent info about this online and have not seen any examples of it within PlasmaPy. Is there some decorator I should use for this? Or would something like `__doc__ += ...` work?

Another good question! I usually end up putting entirely new docstrings in child classes.  This sometimes leads to repetition of the content in child class docstrings, but for the most part, I've found it simpler to have new child class docstrings than to try to add code/logic that modifies (or appends to) the docstrings of the parent classes for the child classes.  Adding in code/logic that modifies docstrings could end up being worth it in more complicated cases (like with a broader range of child classes).

namurphy reviewed (commented):
Thank you again for working on this!  I'm liking how this is coming together!  Here are a few thoughts & suggestions before I head home for the weekend.  I want to spend a bit of time playing around with the classes too, hopefully by next week.  This will be great to have in our next release in September!

namurphy commented on plasmapy/dispersion/analytical/mhd_wave_class.py:
```diff
@@ -0,0 +1,766 @@
+
+class FastMagnetosonicWave(AbstractMHDWave):
+    r"""
+    A class to represent fast magnetosonic waves.
```
I wonder if we'd want to add a note somewhere (maybe in the notes section?) that this is also known as the fast magnetoacoustic wave (and similarly for the slow wave below).  That'd help in case people do a search for that term instead.

TrestanSimon commented on plasmapy/dispersion/analytical/mhd_wave_class.py:
```diff
@@ -0,0 +1,766 @@
+
+    @property
+    def sound_speed(self) -> u.m / u.s:
+        """The sound speed of the plasma."""
```
I think it would be a good idea.

TrestanSimon commented on plasmapy/dispersion/analytical/mhd_wave_class.py:
I was thinking about doing this previously, but I was not sure of what the new name should be. I think this will work.

TrestanSimon commented:
@namurphy Thanks for taking a look at this again. I have implemented all of your suggested changes.

Regarding `mhd_waves` returning a `collections.namedtuple`, I am not sure how to go about writing the new "Returns" section of the docstring. I have attempted to make the change, but I do not think it is correct; could you take a look at it?

Also, since `collections.namedtuple` are indexable, I have changed the applicable tests to loop over the three modes returned in `mhd_waves` by using stacked `@pytest.mark.parametrize` (as you suggest) like so:
```
@pytest.mark.parametrize("mode", range(3))
...
sample_waves[mode].angular_frequency(**kwargs_wave)
```
Utilizing indices seems like the simplest way to go about this, although I am not very familiar with using named tuples and suspect there is a better way. Any thoughts?

ejohnson-96 reviewed (approved):
👍

TrestanSimon commented:
I was able to reproduce the Friedrichs diagrams from Figure 25 of [Huang and Lyu 2019](https://doi.org/10.1063/1.5110991) (with the addition of the Alfvén mode here) using the methods from this PR.

@namurphy You mentioned in the original issue (#2197) maybe having a method to plot Friedrichs diagrams, so I was also wondering if you could expand on this idea.

In the figure below, the Alfvén mode is in green, the fast mode is in red, and the slow mode is in blue.

![Friedrichs_diagrams_HuangLyu19](https://github.com/PlasmaPy/PlasmaPy/assets/40308014/47695371-8f82-4356-b1d9-269914f9f489)

The code I wrote to reproduce these diagrams is in the attached zip file. (GitHub does not allow me to attach py files, and I felt that the code was a little too long to copy and paste here.)
[Friedrichs_diagrams_HuangLyu19.zip](https://github.com/PlasmaPy/PlasmaPy/files/12049469/Friedrichs_diagrams_HuangLyu19.zip)

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+        Parameters
+        ----------
+        k : `~astropy.units.Quantity`, single valued or 1-D array
+            Wavenumber in units convertible to rad/m`.  Either single
+            valued or 1-D array of length :math:`N`.
+        theta : `~astropy.units.Quantity`, single valued or 1-D array
+            The angle of propagation of the wave with respect to the
+            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units must be
+            convertible to radians. Either single valued or 1-D array of
+            size :math:`M`.
```
```suggestion
        k : `~astropy.units.Quantity`
            Wavenumber in units convertible to rad/m`.  Either single
            valued or 1-D array of size :math:`N`.

        theta : `~astropy.units.Quantity`
            The angle of propagation of the wave with respect to the
            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units
            convertible to radians. Either single valued or 1-D array of
            size :math:`M`.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+        -------
+        phase_velocity : `~astropy.units.Quantity`
+            An :math:`N × M` array of computed phase velocities in units
+            m/s.
```
```suggestion
            of m/s.
```

namurphy commented on plasmapy/dispersion/analytical/tests/test_mhd_wave_class.py:
```diff
@@ -0,0 +1,159 @@
+        ],
+    )
+    def test_raises_init(self, kwargs, error):
+        """Test scenarios that raise an `Exception`."""
```
```suggestion
        """Test scenarios that raise an exception."""
```
...since the docstrings for tests don't show up in the docs.

namurphy commented on plasmapy/dispersion/analytical/tests/test_mhd_wave_class.py:
```diff
@@ -0,0 +1,159 @@
+    )
+    @pytest.mark.parametrize("mode", range(3))
+    def test_raises_angular_frequency(self, kwargs, error, mode):
+        """Test scenarios that raise an `Exception`."""
```
```suggestion
        """Test scenarios that raise an exception."""
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+    Parameters
+    ----------
+    B : `~astropy.units.Quantity`
+        The magnetic field magnitude in units convertible to T.
+    density : `~astropy.units.Quantity`
+        Either the ion number density :math:`n_i` in units convertible
+        to m\ :sup:`-3` or the total mass density :math:`ρ` in units
+        convertible to kg m\ :sup:`-3`\ .
+    ion : |particle-like|
+        Representation of the ion species (e.g., ``'p'`` for protons,
+        ``'D+'`` for deuterium, ``'He-4 +1'`` for singly ionized
+        helium-4, etc.). If no charge state information is provided,
+        then the ions are assumed to be singly ionized.
+    T : `~astropy.units.Quantity`, |keyword-only|, optional
+        The plasma temperature in units of K or eV, which defaults
+        to zero.
+    gamma : `float` or `int`, |keyword-only|, optional
+        The adiabatic index for the plasma, which defaults to 3/5.
+    mass_numb : `int`, |keyword-only|, optional
+        The mass number corresponding to ``ion``.
+    Z : `float` or `int`, |keyword-only|, optional
+        The charge number corresponding to ``ion``.
```
```suggestion
    B : `~astropy.units.Quantity`
        The magnetic field magnitude in units convertible to T.

    density : `~astropy.units.Quantity`
        Either the ion number density :math:`n_i` in units convertible
        to m\ :sup:`-3` or the total mass density :math:`ρ` in units
        convertible to kg m\ :sup:`-3`\ .

    ion : |particle-like|
        Representation of the ion species (e.g., ``'p'`` for protons,
        ``'D+'`` for deuterium, ``'He-4 +1'`` for singly ionized
        helium-4, etc.). If no charge state information is provided,
        then the ions are assumed to be singly ionized.

    T : `~astropy.units.Quantity`, |keyword-only|, optional
        The plasma temperature in units of K or eV, which defaults
        to zero.

    gamma : `float` or `int`, |keyword-only|, default: 5/3
        The adiabatic index for the plasma.

    mass_numb : `int`, |keyword-only|, optional
        The mass number corresponding to ``ion``.

    Z : `float` or `int`, |keyword-only|, optional
        The charge number corresponding to ``ion``.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+        Parameters
+        ----------
+        k : `~astropy.units.Quantity`, single valued or 1-D array
+            Wavenumber in units convertible to rad/m`.  Either single
+            valued or 1-D array of length :math:`N`.
+        theta : `~astropy.units.Quantity`, single valued or 1-D array
+            The angle of propagation of the wave with respect to the
+            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units must be
+            convertible to radians. Either single valued or 1-D array of
+            size :math:`M`.
```
```suggestion
        k : `~astropy.units.Quantity`, single valued or 1-D array
            Wavenumber in units convertible to rad/m`. Either single
            valued or 1-D array of length :math:`N`.

        theta : `~astropy.units.Quantity`
            The angle of propagation of the wave with respect to the
            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units
            convertible to radians. Either single valued or 1-D array of
            size :math:`M`.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+    Parameters
+    ----------
+    B : `~astropy.units.Quantity`
+        The magnetic field magnitude in units convertible to T.
+    density : `~astropy.units.Quantity`
+        Either the ion number density :math:`n_i` in units convertible
+        to m\ :sup:`-3` or the total mass density :math:`ρ` in units
+        convertible to kg m\ :sup:`-3`\ .
+    ion : |particle-like|
+        Representation of the ion species (e.g., ``'p'`` for protons,
+        ``'D+'`` for deuterium, ``'He-4 +1'`` for singly ionized
+        helium-4, etc.). If no charge state information is provided,
+        then the ions are assumed to be singly ionized.
+    T : `~astropy.units.Quantity`, |keyword-only|, optional
+        The plasma temperature in units of K or eV, which defaults
+        to zero.
+    gamma : `float` or `int`, |keyword-only|, optional
+        The adiabatic index for the plasma, which defaults to 3/5.
+    mass_numb : `int`, |keyword-only|, optional
+        The mass number corresponding to ``ion``.
+    Z : `float` or `int`, |keyword-only|, optional
+        The charge number corresponding to ``ion``.
```
```suggestion
    B : `~astropy.units.Quantity`
        The magnetic field magnitude in units convertible to T.

    density : `~astropy.units.Quantity`
        Either the ion number density :math:`n_i` in units convertible
        to m\ :sup:`-3` or the total mass density :math:`ρ` in units
        convertible to kg m\ :sup:`-3`\ .

    ion : |particle-like|
        Representation of the ion species (e.g., ``'p'`` for protons,
        ``'D+'`` for deuterium, ``'He-4 +1'`` for singly ionized
        helium-4, etc.). If no charge state information is provided,
        then the ions are assumed to be singly ionized.

    T : `~astropy.units.Quantity`, |keyword-only|, optional
        The plasma temperature in units of K or eV, which defaults
        to zero.

    gamma : `float` or `int`, |keyword-only|, default: 5/3
        The adiabatic index for the plasma.

    mass_numb : `int`, |keyword-only|, optional
        The mass number corresponding to ``ion``.

    Z : `float` or `int`, |keyword-only|, optional
        The charge number corresponding to ``ion``.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+        Parameters
+        ----------
+        k : `~astropy.units.Quantity`, single valued or 1-D array
+            Wavenumber in units convertible to rad/m`.  Either single
+            valued or 1-D array of length :math:`N`.
+        theta : `~astropy.units.Quantity`, single valued or 1-D array
+            The angle of propagation of the wave with respect to the
+            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units must be
+            convertible to radians. Either single valued or 1-D array of
+            size :math:`M`.
```
```suggestion
        k : `~astropy.units.Quantity`
            Wavenumber in units convertible to rad/m`.  Either single
            valued or 1-D array of length :math:`N`.

        theta : `~astropy.units.Quantity`
            The angle of propagation of the wave with respect to the
            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units
            convertible to radians. Either single valued or 1-D array of
            size :math:`M`.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+        .. math::
+
+            ω^2 = \frac{k^2}{2} \left(c_{ms}^2 + \sqrt{c_{ms}^4 - 4 v_A^2 c_s^2 \cos^2 θ}\right)
```
```suggestion
            ω^2 = \frac{k^2}{2} \left(
               c_{ms}^2 + \sqrt{c_{ms}^4 - 4 v_A^2 c_s^2 \cos^2 θ}
            \right)
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+        Parameters
+        ----------
+        k : `~astropy.units.Quantity`, single valued or 1-D array
+            Wavenumber in units convertible to rad/m`.  Either single
+            valued or 1-D array of length :math:`N`.
+        theta : `~astropy.units.Quantity`, single valued or 1-D array
+            The angle of propagation of the wave with respect to the
+            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units must be
+            convertible to radians. Either single valued or 1-D array of
+            size :math:`M`.
```
```suggestion
        k : `~astropy.units.Quantity`
            Wavenumber in units convertible to rad/m`. Either single
            valued or 1-D array of length :math:`N`.

        theta : `~astropy.units.Quantity`
            The angle of propagation of the wave with respect to the
            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units
            convertible to radians. Either single valued or 1-D array of
            size :math:`M`.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+        Parameters
+        ----------
+        k : `~astropy.units.Quantity`, single valued or 1-D array
+            Wavenumber in units convertible to rad/m`.  Either single
+            valued or 1-D array of length :math:`N`.
+        theta : `~astropy.units.Quantity`, single valued or 1-D array
+            The angle of propagation of the wave with respect to the
+            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units must be
+            convertible to radians. Either single valued or 1-D array of
+            size :math:`M`.
```
```suggestion
        k : `~astropy.units.Quantity`
            Wavenumber in units convertible to rad/m`. Either single
            valued or 1-D array of length :math:`N`.

        theta : `~astropy.units.Quantity`
            The angle of propagation of the wave with respect to the
            magnetic field, :math:`\cos^{-1}(k_z / k)`, in units
            convertible to radians. Either single valued or 1-D array of
            size :math:`M`.
```

namurphy commented on plasmapy/dispersion/analytical/mhd_waves_.py:
```diff
@@ -0,0 +1,1078 @@
+
+    Parameters
+    ----------
+    B : `~astropy.units.Quantity`
+        The magnetic field magnitude in units convertible to T.
+    density : `~astropy.units.Quantity`
+        Either the ion number density :math:`n_i` in units convertible
+        to m\ :sup:`-3` or the total mass density :math:`ρ` in units
+        convertible to kg m\ :sup:`-3`\ .
+    ion : |particle-like|
+        Representation of the ion species (e.g., ``'p'`` for protons,
+        ``'D+'`` for deuterium, ``'He-4 +1'`` for singly ionized
+        helium-4, etc.). If no charge state information is provided,
+        then the ions are assumed to be singly ionized.
+    T : `~astropy.units.Quantity`, |keyword-only|, default: 0 K
+        The plasma temperature in units of K or eV.
+    gamma : `float` or `int`, |keyword-only|, optional
+        The adiabatic index for the plasma, which defaults to 3/5.
+    mass_numb : 'int', |keyword-only|, optional
+        The mass number corresponding to ``ion``.
+    Z : `float` or 'int', |keyword-only|, optional
+        The charge number corresponding to ``ion``.
```
```suggestion
    B : `~astropy.units.Quantity`
        The magnetic field magnitude in units convertible to T.

    density : `~astropy.units.Quantity`
        Either the ion number density :math:`n_i` in units convertible
        to m\ :sup:`-3` or the total mass density :math:`ρ` in units
        convertible to kg m\ :sup:`-3`\ .

    ion : |particle-like|
        Representation of the ion species (e.g., ``'p'`` for protons,
        ``'D+'`` for deuterium, ``'He-4 +1'`` for singly ionized
        helium-4, etc.). If no charge state information is provided,
        then the ions are assumed to be singly ionized.

    T : `~astropy.units.Quantity`, |keyword-only|, default: 0 K
        The plasma temperature in units of K or eV.

    gamma : `float` or `int`, |keyword-only|, default: 5/3
        The adiabatic index for the plasma.

    mass_numb : 'int', |keyword-only|, optional
        The mass number corresponding to ``ion``.

    Z : `float` or 'int', |keyword-only|, optional
        The charge number corresponding to ``ion``.
```

namurphy reviewed (approved):
Apologies for the delay in this review!  I really like how it turned out.  I made a few minor suggestions (mostly for consistency with other docstrings) which I'll accept and then merge if tests pass so we can get it out in the next release.  Thank you again so much!
